pkg/metering/writer.go

```diff
@@ -0,0 +1,191 @@
+// Copyright 2025 TiKV Project Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metering
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"go.uber.org/zap"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/log"
+	"github.com/pingcap/metering_sdk/common"
+	"github.com/pingcap/metering_sdk/config"
+	"github.com/pingcap/metering_sdk/storage"
+	meteringwriter "github.com/pingcap/metering_sdk/writer/metering"
+
+	"github.com/tikv/pd/pkg/utils/logutil"
+)
+
+// meteringFlushInterval is the interval to flush the metering to the underlying storage.
+const meteringFlushInterval = time.Minute
+
+// Collector collects events into records with caller-defined fields.
+type Collector interface {
+	Category() string
+	Collect(data any)
+	Flush() []map[string]any
+}
+
+// Writer is used to as a delegate to collect and flush the metering data to the underlying storage.
+type Writer struct {
+	sync.RWMutex
+	id       string
+	ctx      context.Context
+	cancel   context.CancelFunc
+	wg       sync.WaitGroup
+	provider storage.ObjectStorageProvider
+	inner    *meteringwriter.MeteringWriter
+	// collectors holds category -> collector, allowing callers to customize record fields per category.
+	collectors map[string]Collector
+}
+
+// NewWriter creates a new metering writer to collect and report the metering data to the underlying storage.
+func NewWriter(ctx context.Context, meteringConfig *Config, id string) (*Writer, error) {
+	err := meteringConfig.adjust()
+	if err != nil {
+		return nil, err
+	}
+
+	// Create storage provider
+	providerConfig := &storage.ProviderConfig{
+		Type:   meteringConfig.Type,
+		Bucket: meteringConfig.Bucket,
+		Region: meteringConfig.Region,
+		Prefix: meteringConfig.Prefix,
+	}
+	provider, err := storage.NewObjectStorageProvider(providerConfig)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create storage provider for metering")
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	return &Writer{
+		id:       id,
+		ctx:      ctx,
+		cancel:   cancel,
+		provider: provider,
+		inner: meteringwriter.NewMeteringWriter(
+			provider,
+			config.DefaultConfig().WithLogger(zap.L()),
+		),
+		collectors: make(map[string]Collector),
+	}, nil
+}
+
+// Start starts the metering writer background loop
+func (mw *Writer) Start() {
+	if mw == nil {
+		return
+	}
+	mw.wg.Add(1)
+	go mw.meteringLoop()
+	log.Info("metering writer started")
+}
+
+// Stop stops the metering writer
+func (mw *Writer) Stop() {
```

Comment:
The metering data after the last ticker is not recorded, is it necessary to flush before stopping?